feat(rules): support prefixes in config

supermaxiste opened the pull request:
## Proposed Changes

A new parsing step for tripsu rules to allow the use of prefixes in a config. Addresses #60 

## Types of Changes

What types of changes does your code introduce? _Put an `x` in the boxes that
apply_

- [ ] A **bug fix** (non-breaking change which fixes an issue).
- [x] A new **feature** (non-breaking change which adds functionality).
- [ ] A **breaking change** (fix or feature that would cause existing
      functionality to not work as expected).
- [ ] A **non-productive** update (documentation, tooling, etc. if none of the
      other choices apply).

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [x] I have read the
      [CONTRIBUTING](https://github.com/sdsc-ordes/tripsu/blob/master/CONTRIBUTING.md)
      guidelines.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added the necessary documentation (if appropriate).

## Further Comments

These are 2 new main functions that are added as `Impl` to `Rules` in `rules.rs`:

1. `has_valid_curies_and_uris` which checks the content of the rules to find any mistake or problem
2. `expand_rules_curie` which takes rules with or without cURIs and returns a set of rules with full URIs

Each one of the main functions comes with their specific set of tests.

cmdoret commented on src/rules.rs:
```diff
@@ -30,13 +32,250 @@ pub struct Rules {
+    pub fn has_valid_curies_and_uris(&self) -> bool {
+        match &self.prefix {
+            // If no prefix are set, check each URI for validity
+            None => self.check_uris(&self.nodes, &self.objects),
```
**suggestion:** as `self.check_uris` already receives `self`, we don't need to pass the specific attributes, they can be accessed on `self` inside the body.

cmdoret reviewed (changes requested):
Great feature addition!

I have added many suggestions, but broadly they fall into two categories:
* scoping: instead of pooling all the logic into `Rules`, we should leverage [composition](https://refactoring.guru/design-patterns/composite) and modules to keep functions in the right scope. This improve [cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) and makes the logic easier to read thanks to modularity.
* Error handling, using `Result` in function signatures would reduce the amount of conditions checking.

supermaxiste commented on src/crypto.rs:
```diff
@@ -5,7 +5,7 @@ use rand::Rng;
 pub(crate) fn generate_key(size: usize) -> Vec<u8> {
     let mut rng = rand::thread_rng();
-    return (0..size).map(|_| rng.gen::<u8>()).collect();
+    (0..size).map(|_| rng.gen::<u8>()).collect()
```
Thank you for the heads up! I didn't know about the option to `just lint` and I think I interpreted the CI wrong. I also don't mind following Rust idioms ;) 

I will also make sure to better separate linting/formatting commits in the future

supermaxiste commented on src/pseudo.rs:
```diff
@@ -94,10 +94,10 @@ mod tests {
-            &input_path,
-            &rules_path,
+            input_path,
+            rules_path,
             &output_path,
-            &type_map_path,
+            type_map_path,
             &key,
```
On my rust analyzer all of these input paths are already references, which is why they're passed as is. The `Path::new` function returns a reference. A more elaborate answer [here](https://stackoverflow.com/questions/79090075/trying-to-understand-rusts-magic-in-stdpathpathas-ref).

supermaxiste commented on tests/data/rules.yaml:
```diff
@@ -1,21 +1,26 @@
 invert: false
 
+# Define prefixes for the URIs used in the rules.
+prefix:
```
Addressed

supermaxiste commented on src/rules.rs:
```diff
@@ -30,13 +32,250 @@ pub struct Rules {
+        // We assume that a full URI starts with "<" and ends with ">"
+        // We select the URI within the angle brackets
+        Iri::new(&uri[1..uri.len() - 2]).is_ok()
+    }
```
Added new module `uris.rs` and moved `check_string_iri` there

supermaxiste commented on src/rules.rs:
```diff
@@ -30,13 +32,250 @@ pub struct Rules {
+        Iri::new(&uri[1..uri.len() - 2]).is_ok()
+    }
+
+    fn filter(&self, hash_set: &HashSet<String>) -> HashSet<String> {
```
Renamed to `filter_out_full_uris` and moved to `uris` module

supermaxiste commented on src/rules.rs:
```diff
@@ -30,13 +32,250 @@ pub struct Rules {
+    fn is_full_uri(&self, uri: &str) -> bool {
+        // Ensure that full URI starts with "<" and ends with ">"
+        uri.starts_with('<') && uri.ends_with('>')
+    }
```
done

supermaxiste commented on src/rules.rs:
```diff
@@ -30,13 +32,250 @@ pub struct Rules {
+                self.check_string_iri(&k) && v.into_iter().all(|uri| self.check_string_iri(&uri))
+            })
+    }
+    fn check_string_iri(&self, uri: &str) -> bool {
```
done

cmdoret replied:
:warning: @supermaxiste looks like you forgot to add `uri.rs` into git :warning:

supermaxiste commented:
# Major

- Composition and cohesion improved by:
  - Adding a new `uris.rs` module taking care of all operations related to URIs
  - Adding `Node`  and `Object`  specific `impl`  for operations on them
  - Creating a `CompactUriSet` and `PrefixMap` structs within the URI module to better handle cURIs and prefix maps
- Error handling
    - Mistakes in both URIs and cURIs are now better error handled (see examples below)
- Default prefix support
  - the `default` key in the `prefixes` field can now be used as default prefix for all cURIs without a prefix in front of them (see examples below)
  - added tests for valid and invalid default prefixes 

# Minor

- Improved function naming
- Added descriptions to functions

# Examples

### A valid rule set using default

```yaml
prefixes:
  default: "<http://example.org/>"
  foaf: "<http://xmlns.com/foaf/0.1/>"

nodes:
  of_type:
  - "Person"
```

### An invalid ruleset not providing a default but using one

```yaml
prefixes:
  foaf: "<http://xmlns.com/foaf/0.1/>"

nodes:
  of_type:
  - "Person"
```

`Error expanding rules curie. No default prefix provided for: Person`

### An invalid ruleset with a typo on full URIs

```yaml
nodes:
  of_type:
  - "<http://xmlns.com/foaf/0.1/Person"
```
`Error expanding rules curie. Invalid prefix: <http://xmlns.com/foaf/0.1/Person`

### An invalid ruleset with a cURI but no prefix for it

```yaml
nodes:
  of_type:
  - "ex:Person"
```
`Error expanding rules curie. Invalid prefix: ex:Person`

cmdoret commented:
Great news! :smiley: 
Just reading the spec before reviewing the code:
* cURIes with the default prefix are normally denoted as `:Person` and not `Person`. We may want to stick with that.
* Similarly, in other RDF serialization format, the default prefix is simply the empty string, there is no special name.

With these two together, and yaml supporting the empty string as a key, we could have:

```yaml
prefixes:
  "": "<http://example.org/>"
  foaf: "<http://xmlns.com/foaf/0.1/>"

nodes:
  of_type:
  - ":Person"
  - "foaf:Friend"
```

IIUC, this means we do not need to have any special code / exception for the default prefix; it's just a normal prefix using the  empty string.

supermaxiste commented:
> CURIES with the default prefix are often denoted as `:Person` and not `Person`.

I double checked the `curie` crate and both `:Person` and `Person` [are supported](https://docs.rs/curie/latest/curie/). Should I add this somewhere in the README or tutorial?

> Similarly, in most RDF serialization formats, the default prefix is simply the empty string, there is no special name.

We can simply change [this line](https://github.com/sdsc-ordes/tripsu/blob/ff57bda322a1c6982670dbd7c90c2371c01238ae/src/uris.rs#L46).

supermaxiste commented:
Added what you requested with some important points. 

## How the curie crate handles cURIs
I realized that the `curie` crate handles `Person` and `:Person` differently:

- `Person` can be expanded ONLY with a `default` prefix key
- `:Person` can be expanded ONLY with a `""` prefix key

Because of this reason we're now using only `:Entries` in our rules.

## Minor but important test change

I had to change a valid default test to explicitly include the `""` characters to set a default prefix:

- Before: 
```rust
#[case("", "<http://example.org/>", ":Person", "<http:hasName>", true)]
```
- After: 
```rust
#[case("\"\"", "<http://example.org/>", ":Person", "<http:hasName>", true)]
```

## Handling "" in the prefixes with the `Rules` struct

Previously we defined prefixes as a HashMap of strings:
```rust
prefixes: Option<HashMap<String, String>>
```
But this leads to a parsing error when dealing with an empty key for default. Now the first string is set as optional:

```rust
prefixes: Option<HashMap<Option<String>, String>>
```

This change only affects the `uris.rs` module, specifically the `impl` that imports the serde deserialisation into a prefix map (`import_hashmap`). The function now handles the default as follows:

```rust
// If there's a prefix, add it as expected
if let Some(prefix) = key.as_deref() {
                    if let Err(e) = self.0.add_prefix(prefix, &value[1..value.len() - 1]) {
                        eprintln!("Failed to add prefix: {:?}", e);
                    }
// If there's nothing, set a default prefix
                } else {
                    self.0.set_default(&value[1..value.len() - 1])
                }
```

cmdoret reviewed (approved):
LGTM :)) nice work! :raised_hands:
